app/streak/FilterBuilderInput.tsx

```diff
@@ -0,0 +1,351 @@
+'use client';
+import {
+  useEffect,
+  useRef,
+  useState,
+  type ChangeEventHandler,
+  type KeyboardEventHandler,
+} from 'react';
+import fuzzy from 'fuzzy';
+import { type AllRelationalOperators, type LogicalOperator } from './operators';
+import { FilterPill } from './FilterPill';
+
+export type FilterValue = string | Date | boolean;
+export type FullFilterSelectionFn = (newFilter: FullFilter) => void;
+export type FilterValueSelectionFn = (arg: FilterValue[]) => void;
+export type FilterSelectionComponentProps = {
+  input?: string;
+  onFilterValueSelect?: FilterValueSelectionFn;
+};
+export type FilterDefinition = {
+  filterKey: string;
+  validRelationalOperators: readonly AllRelationalOperators[];
+  options?: Record<string, any>[];
+  optionKey: string;
+  getOptions?:
+    | ((input: string) => Record<string, any>[])
+    | ((input: string) => Promise<Record<string, any>[]>);
+  shouldAllowCustomText: boolean;
+};
+
+// onFilterUpdate
+export type FullFilter = {
+  localId: string;
+  filterKey: string;
+  values: FilterValue[]; // will probably only accept single value for first iteration
+  logicalOperator: LogicalOperator;
+  relationalOperator: string;
+};
+export type FilterState = Array<FullFilter>;
+export type FilterUpdateFunction = ({
+  newFilterState,
+}: {
+  newFilterState: FilterState;
+}) => Promise<void>;
+
+export type FilterBuilderArgs = {
+  filterState: FilterState;
+  onFilterUpdate: FilterUpdateFunction;
+  filterDefinitions: readonly FilterDefinition[];
+};
+
+export const allowedInputBuiltinTypes = [
+  'button',
+  'checkbox',
+  'color',
+  'date',
+  'email',
+  'hidden',
+  'number',
+  'search',
+  'text',
+];
+export const FilterBuilderInput: React.FC<FilterBuilderArgs> = ({
+  filterDefinitions,
+  filterState,
+  onFilterUpdate,
+}) => {
+  const [currFilter, setCurrFilter] = useState<FilterDefinition | null>(null);
+  const [currRelOperator, setCurrRelationalOperator] = useState<string | null>(
+    null,
+  );
```

Comment:
`useReducer` might be better here to handle the state with actions for this component.